service/src/main/java/bio/terra/workspace/service/admin/flights/cloudcontexts/gcp/GcpIamCustomRolePatchStep.java

```diff
@@ -0,0 +1,106 @@
+import com.google.api.client.googleapis.json.GoogleJsonResponseException;
+import com.google.api.services.iam.v1.model.Role;
+import java.io.IOException;
+import java.util.HashSet;
+
+/** Step to update IAM custom project and resource roles in GCP project. */
+public class GcpIamCustomRolePatchStep implements Step {
+
+  private final IamCow iamCow;
+  private final String projectId;
+
+  private final HashSet<CustomGcpIamRole> customGcpIamRoles = new HashSet<>();
+
+  public GcpIamCustomRolePatchStep(IamCow iamCow, String projectId) {
+    this.iamCow = iamCow;
+    this.projectId = projectId;
+    customGcpIamRoles.addAll(CUSTOM_GCP_IAM_ROLES);
+    customGcpIamRoles.addAll(CustomGcpIamRoleMapping.CUSTOM_GCP_RESOURCE_IAM_ROLES.values());
+  }
+
+  @Override
+  public StepResult doStep(FlightContext flightContext)
+      throws InterruptedException, RetryException {
+    for (CustomGcpIamRole customProjectRole : customGcpIamRoles) {
+      updateCustomRole(customProjectRole, projectId);
+    }
+    return StepResult.getStepResultSuccess();
+  }
+
+  /**
+   * Utility for creating custom roles in GCP from WSM's CustomGcpIamRole objects. These roles will
+   * be defined at the project level specified by projectId.
+   */
+  private void updateCustomRole(CustomGcpIamRole customRole, String projectId)
+      throws RetryException {
+    String fullyQualifiedRoleName = customRole.getFullyQualifiedRoleName(projectId);
+    try {
+      Role role = iamCow.projects().roles().get(fullyQualifiedRoleName).execute();
+      role.setIncludedPermissions(customRole.getIncludedPermissions());
+      iamCow.projects().roles().patch(fullyQualifiedRoleName, role).execute();
+    } catch (IOException e) {
+      handleIOException(e);
+    }
+  }
+
+  @Override
+  public StepResult undoStep(FlightContext flightContext) throws InterruptedException {
+
+    for (CustomGcpIamRole customGcpIamRole : customGcpIamRoles) {
+      resetCustomRolesToPreviousSetOfPermissions(flightContext, customGcpIamRole);
+    }
+    return StepResult.getStepResultSuccess();
+  }
+
+  private void resetCustomRolesToPreviousSetOfPermissions(
+      FlightContext flightContext, CustomGcpIamRole customGcpIamRoles) {
+    FlightMap workingMap = flightContext.getWorkingMap();
+    Role originalRole =
+        workingMap.get(customGcpIamRoles.getFullyQualifiedRoleName(projectId), Role.class);
+    if (originalRole != null) {
+      try {
+        iamCow
+            .projects()
+            .roles()
+            .patch(
+                originalRole.getName(),
+                new Role().setIncludedPermissions(originalRole.getIncludedPermissions()))
+            .execute();
+      } catch (IOException e) {
+        handleIOException(e);
+      }
+    } else {
+      try {
+        iamCow
+            .projects()
+            .roles()
+            .delete(customGcpIamRoles.getFullyQualifiedRoleName(projectId))
```

Comment:
Will the `patch` call in the do step create the role if it doesn't exist? If not, I think undo shouldn't delete it, since this flight couldn't have created it.

Reply:
oh good catch.. i was mid-implementing this part and forgot..

i think we should create a new role if it didn't exist.